Add a section on encoding IBCReceive errors

ethanfrey opened the pull request:
Builds on #898 (merge that first)

Closes #762 

- [x] Document desired behavior
- [x] Provide code sample of how to use it

ethanfrey commented:
I'd be happy for a review on this one. Especially from @alpe who has seen the sdk bindings.

I hope this provides a good solution for rollback on receive

ethanfrey commented:
Open question: do we want to do something similar with acknowledgement and timeout?

webmaster128 commented on IBC.md:
```diff
@@ -265,24 +265,27 @@ pub fn ibc_packet_receive(
+understand it fully).
+
+Also note the different return response here (`IbcReceiveResponse` rather than
 `IbcBasicResponse`)? This is because it has an extra field
```
```suggestion
`IbcBasicResponse`). This is because it has an extra field
```

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+
+1. We only return `Ok(Response)` or `Err(String)`. There is no way to return an
+   arbitrary binary encoded message in the error variant.
+2. If we return `messages` to be dispatched, they may fail with an error, which
+   is returned directly to the x/wasm handler. Obviously an error from the bank
+   handler will not be in the proper IBC Acknowledgement format.
```
Is this paragraph about messages or sub-messages? Because for messages the transaction is terminated and reverted, such that no acknowledgement can be created, no matter if the format is known.

ethanfrey replied:
> Because for messages the transaction is terminated and reverted, such that no acknowledgement can be created, no matter if the format is known.

Oh, this will change for this one entry point. That is the crux of this proposal. An error from a message will be handled like an error from `ibc_receive_packet`, which means it will be sent to the encoder.

Please let me know how I can express this clearer.

webmaster128 replied:
I would not do this. This changes fundamental principles of how messages work. What's preventing `ibc_receive_packet` implementors to use sub messages instead of messages?

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+   not even have the information needed to format the success response, so we
+   will need some delayed response-builder in this case also.
+
+eg. In the ICS27 spec, they define a packet to register a new account, the
```
```suggestion
Eg. in the ICS27 spec, they define a packet to register a new account, the
```

webmaster128 replied:
Can we find a nicer way of saying "they"? E.g. "the authors", "it is defined"

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+   changes
+2. Call `ibc_receive_packet`, get either `Err(String)` or
+   `Ok(Response + acknowledgement)`. If `Err`, jump to 4.
+3. Execute any `submessages` and `messages` returned with this response. If any
+   return `Err`, jump to 4.
```
Do we really want to change how messages behave? I'd avoid that

alpe replied:
This is also not clear to me. Why do we need step 4 for messages? Submessages are available instead.

ethanfrey replied:
`on_packet_receive` may return an error (revert contract state)

`OnPacketReceive` in `x/wasm` should *never* return an error, but rather revert the contract state and return binary data to the sender to explain it failed.

If my contract sends tokens, but for some reason that fails (insufficient funds), it should not abort the tx (so the same packet is posted again and again), but make no state change and return an error acknowledgement.

I suggest reading through the 3 linked cosmos-sdk/ibc-go issues/PRs and all the comments for full context. (I know it is long, but this is the approach they are going for in the sdk and what I want to support in CosmWasm)

ethanfrey replied:
In go terms, the `ibc_encode_acknowledgement` works like a `defer()` on  the result

ethanfrey replied:
If you understand what I am trying to explain I would be grateful for a better way to express it

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+    env: Env,
+    // the original packet we got if we need some data from that
+    packet: IbcPacket,
+    // Either Ok(ibc_receive_packet acknowledgement) or Err(failure message)
+    receiveResult: Result<Binary, String>,
```
I'd split those to allow accessing the "receive data" in the error case as well:

```suggestion
    // Data returned by `ibc_receive_packet`
    receive_data: Binary,
    // Either Ok(()) or Err(failure message)
    processing_result: Result<(), String>,
```

ethanfrey replied:
Okay, so `ibc_receive_packet` may return `acknowledgement: Binary` or may error.

If it succeeded but a dispatched (sub-)message errored, then we have both an error (eg. from BankMsg) and an acknowledgement (from the original call).

We end up with:

```
receive_data: Option<Binary>,
processing_result: Result<(), String>,
```

Does that make sense?

webmaster128 replied:
I would not call the ack encoder/creator entry point if `ibc_receive_packet` errors but instead fail the transaction execution (same discussions as with messages).

Then `receive_data` is always set (to whatever `ibc_receive_packet` returned).

ethanfrey replied:
Continuing on discord. I have clearly failed to explain the IBC architecture design that we fit into.

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+are passed just in case more context is needed for encoding, but likely unused.
+This function should never return an error (even if the IbcPacket was
+malformed), but we allow it to return one if there is some pathological state
+(rather than panicking in the contract,`StdResult::Err` returns useful
```
```suggestion
(rather than panicking in the contract, `StdResult::Err` returns useful
```

webmaster128 commented on IBC.md:
```diff
@@ -330,12 +424,18 @@ message Acknowledgement {
 whether to encode it as JSON or Protobuf. In the ICS20 implementation, this is
-JSON encoded when returned from a contract. Given that, we will consider this
-structure, JSON-encoded, to be the "standard" acknowledgement format.
+JSON encoded when returned from a contract. In ICS27, they are discussing using
```
```suggestion
JSON encoded when returned from a contract. In ICS27, the authors are discussing using
```

alpe commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+4. `receiveResult` is `Err(String)` if any of the above errored, or `Ok(Binary)`
+   with the original acknowledgement bytes if we have the success case.
+5. Call `ibc_encode_acknowledgement` with the `receiveResult`. The return value
+   is the acknowledgement packet to send to the calling chain. This should never
```
What is the DB transaction scope of this call? Is it the same as `ibc_receive_packet` so that it would be reverted on the error before or a new one that is independent?
I don't have a concrete use case but it feels useful to persist data with a failure ACK as well. If we want this then an error return value makes sense. Though the process would end without an ACK to the packet sender.

ethanfrey replied:
This is not my invention, I am following the ibc design (which will come out in 0.43 or 0.44). Check the links at the top of https://github.com/CosmWasm/cosmwasm/blob/on-packet-recv-error-handling/IBC.md#acknowledgement-processing

Especially read the issue description here: https://github.com/cosmos/ibc-go/issues/91 (skim the comments if you wish, the description has the intention, the rest is discussing details). And if you want to dig into the code to understand one part, you can see https://github.com/cosmos/ibc-go/pull/107 (but that is not necessary to get the design)

alpe commented on IBC.md:
```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+   is the acknowledgement packet to send to the calling chain. This should never
+   error.
+6. If `ReceiveResult` is `Ok`, commit this "storage cache". If it is `Err`,
+   revert those changes.
```
🤔 Do I understand this correct that this step let the contract handle the error case on messages and control the TX scope? The process I was assuming would revert the TX on error in 3) but can persist new state in 5).

ethanfrey replied:
The future IBC design (linked above, here again: https://github.com/cosmos/ibc-go/issues/91) explicitly makes the entire app execution (x/wasm) atomic - to be rolled back when app error. But not touching the IBC part of the transaction and successfully committing a "error acknowledgement" in the IBC Keeper.

I feel it is pointless to design something other than what they are doing, unless you want to re-open that issue in the ibc repo. As in the mid-term, we will just use those semantics in `x/wasm` and all other modules will use them as well.

ethanfrey commented:
Okay, this got a complete overhaul since I worked out the new `SEMANTICS.ms`. Please take a fresh look at this.

I also modified `ibc-reflect` to use this technique in `ibc_packet_receive` (as well as submessages in `ibc_channel_connect` to demo normal "callback" usage).

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+we struggled to map this idea to the CosmWasm model. However, we also discovered
+a deep similarity between these requirements and the
+[submessage semantics](./SEMANTICS.md#submessages). It just requires some
+careful coding on the contract developer's side to new throw errors. This
```
Is this what you want to say?

```suggestion
careful coding on the contract developer's side to not throw errors. This
```

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+a deep similarity between these requirements and the
+[submessage semantics](./SEMANTICS.md#submessages). It just requires some
+careful coding on the contract developer's side to new throw errors. This
+produced 3 suggests on how to handle errors and rollbacks _inside
```
```suggestion
produced 3 suggestions on how to handle errors and rollbacks _inside
```

webmaster128 commented on IBC.md:
````diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+   acknowledgements. This would look something like the
+   [main dispatch loop in `ibc-reflect`](https://github.com/CosmWasm/cosmwasm/blob/cd784cd1148ee395574f3e564f102d0d7b5adcc3/contracts/ibc-reflect/src/contract.rs#L217-L248):
+
+```rust
````
You can indent it such that the whole code block belongs to 1., not below 1.:

```suggestion
   ```rust
```

webmaster128 commented on IBC.md:
````diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+            attributes: vec![],
+        })
+    })
+```
````
We could even enforece this by changing the signature of the entry point to `IbcReceiveResponse`, i.e. no result around it.

ethanfrey replied:
Nice idea. This will push devs to proper usage

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+   [bottom of reply section](./SEMANTICS.md#handling-the-reply)). You can see a
+   similar example in how
+   [`ibc-reflect` handles `receive_dispatch`](https://github.com/CosmWasm/cosmwasm/blob/eebb9395ccf315320e3f2fcc526ee76788f89174/contracts/ibc-reflect/src/contract.rs#L307-L336).
+   Note how we use a unique reply id for this and use that to catch any
```
```suggestion
   Note how we use a unique reply ID for this and use that to catch any
```

webmaster128 commented on IBC.md:
````diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+   Note how we use a unique reply id for this and use that to catch any
+   execution failure and return an error acknowledgement instead:
+
+```rust
````
```suggestion
   ```rust
```

webmaster128 commented on IBC.md:
```diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
+         ..Response::default()
+      }),
+      (INIT_CALLBACK_ID, ContractResult::Ok(response)) => handle_init_callback(deps, response),
+      _ => Err(StdError::generic_err("invalid reply id")),
```
```suggestion
      _ => Err(StdError::generic_err("invalid reply id or result")),
```

webmaster128 commented on contracts/ibc-reflect/src/contract.rs:
```diff
@@ -36,31 +38,42 @@ pub fn instantiate(
+            ..Response::default()
+        }),
+        (INIT_CALLBACK_ID, ContractResult::Ok(response)) => handle_init_callback(deps, response),
+        _ => Err(StdError::generic_err("invalid reply id")),
```
```suggestion
        _ => Err(StdError::generic_err("invalid reply id or result")),
```

webmaster128 reviewed (commented):
Yeah, very helpful!

ethanfrey commented on IBC.md:
```diff
@@ -265,24 +265,27 @@ pub fn ibc_packet_receive(
+contains all information needed to process the receipt. You can generally ignore
+timeout (this entry point is only called if it hasn't yet timed out) and
+sequence (which is used by the IBC framework to avoid duplicates). I generally
+use `dest.channel_id` like `info.sender` to authenticate the packet, and parse
```
Yes, this info comes from the ibc module after it is validated. Only when everything is verified does the contract receive the various `ibc_*` packets. I should make this more clear. `packet IbcPacket` is trusted data, just like `info MessageInfo`.

ethanfrey commented:
I addressed all the comments except for changing the return type. If we want to do that, I would like to do that in a different PR (this one is mainly docs and making example code).

Can we merge this?
Shall I start a new PR on top with the return value change?

alpe commented on IBC.md:
```diff
@@ -330,12 +480,18 @@ message Acknowledgement {
+Note that it leaves the actual success response as app-specific bytes where you
+can place anything, but does provide a standard way for an observer to check
+success-or-error. If you are designing a new protocol, I encourage you to use
+this struct in either of the encodings as the acknowledgement envelope.
```
👍  good to have this as recommendation

alpe reviewed (approved):
Thanks for the writeup. This makes sense 👍
